{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d3b69ef3-5a82-4dc5-91da-08d45d4bfb99",
   "metadata": {
    "id": "827a7a6e"
   },
   "source": [
    "# Experiment: Use classic machine learning approach (Random Forest) for relation classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "348c3772-f43f-4b7f-84d3-43c9f362372b",
   "metadata": {
    "id": "827a7a6e"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\fschr\\AppData\\Roaming\\Python\\Python311\\site-packages\\pandas\\core\\arrays\\masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).\n",
      "  from pandas.core import (\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from sklearn.metrics import classification_report\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "pd.set_option(\"max_colWidth\", 300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fda3198f-bc3d-4042-9bbd-57db8c3d7649",
   "metadata": {
    "id": "73b1aa82"
   },
   "source": [
    "# Read the datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "26fb0bc9-4fae-4731-891a-782874dfd264",
   "metadata": {
    "id": "W0DEbJkLg4Y6"
   },
   "outputs": [],
   "source": [
    "train_set = pd.read_json(\"../../data/subtask3/train_prepro.json\")\n",
    "valid_set = pd.read_json(\"../../data/subtask3/valid_prepro.json\")\n",
    "test_set = pd.read_json(\"../../data/subtask3/test_prepro.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "03cf6acd-db82-4acb-ae6f-1e457dc6397e",
   "metadata": {
    "id": "W0DEbJkLg4Y6"
   },
   "outputs": [],
   "source": [
    "labels = [\n",
    "    'Version_of',\n",
    "    'Developer_of',\n",
    "    'Citation_of',\n",
    "    'URL_of',\n",
    "    'PlugIn_of',\n",
    "    'Specification_of',\n",
    "    'Release_of',\n",
    "    'Extension_of',\n",
    "    'License_of',    \n",
    "    'Abbreviation_of',\n",
    "    'AlternativeName_of',\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d98ab68-97c7-4228-8f52-88ec076b494d",
   "metadata": {
    "id": "W0DEbJkLg4Y6"
   },
   "source": [
    "# Feature Selection and Preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "59e09bfb-d12e-4bee-aa2e-1bbfd8d4cac0",
   "metadata": {
    "id": "W0DEbJkLg4Y6"
   },
   "outputs": [],
   "source": [
    "features_text = [\n",
    "      # 'context',\n",
    "       'sentence',\n",
    "       'subject_text',\n",
    "       'object_text'\n",
    "]\n",
    "features_cat = [\n",
    "       #'target_label_binary',\n",
    "       'subject_label',\n",
    "       'subject_intention',\n",
    "       'object_label',\n",
    "       'object_intention',\n",
    "       # Maybe other labels? 'entities',        \n",
    "]\n",
    "features_int = [\n",
    "    \"subject_begin\",\n",
    "    \"subject_end\",\n",
    "    \"object_begin\",\n",
    "    \"object_end\",\n",
    "    'subject_object_distance',\n",
    "    'max_distance_train',\n",
    "    'left_to_right', # binary\n",
    "    'unprobable_relation_direction', # binary\n",
    "    'n_alternative_candidates_by_subject',\n",
    "    'n_alternative_candidates_by_subject_filtered',\n",
    "    'rank_by_candidate_distance', 'rank_by_candidate_distance_filtered'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "40abc079-47d9-41dd-beb0-dd15445df2d5",
   "metadata": {
    "id": "W0DEbJkLg4Y6"
   },
   "outputs": [],
   "source": [
    "## Not giving better results (kept, maybe helpfull later)\n",
    "def text_between(rel):\n",
    "    token = rel[\"sentence\"].split()\n",
    "    if rel.subject_end < rel.object_begin:\n",
    "        return \" \".join(token[rel.subject_end + 1: rel.object_begin])\n",
    "    if rel.object_end < rel.subject_begin:\n",
    "        return \" \".join(token[rel.object_end + 1 : rel.subject_begin])\n",
    "train_set[\"text_between_ents\"] = train_set[[\"sentence\", \"subject_begin\", \"subject_end\", \"object_begin\", \"object_end\"]].apply(text_between, axis=1)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2301b0b0-f25f-4400-a698-e87973ea49b4",
   "metadata": {
    "id": "W0DEbJkLg4Y6"
   },
   "source": [
    "### Build tokenizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c7663ae3-8a2f-4919-b3b2-3ce3d1038ace",
   "metadata": {
    "id": "W0DEbJkLg4Y6"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((1150, 1183), (4450, 1275), (636, 808))"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "punctuation = \"\\.\\(\\)\\[\\]\\,\\:\\;\\\"\\'\\/_/+\"\n",
    "token_pattern = f'(?u)(?:\\\\b\\\\w+\\\\b|[{punctuation}])'\n",
    "vect = TfidfVectorizer(min_df=10, ngram_range=(1,2), token_pattern=token_pattern)\n",
    "vect_subj = TfidfVectorizer(min_df=1, analyzer=\"word\", ngram_range=(1,1), token_pattern=token_pattern, lowercase=False)\n",
    "vect_obj = TfidfVectorizer(min_df=1, analyzer=\"word\", ngram_range=(1,1), token_pattern=token_pattern, lowercase=False)\n",
    "hih = vect.fit_transform(train_set.text_between_ents)\n",
    "hah = vect_subj.fit_transform(train_set.subject_text.unique())\n",
    "huh = vect_obj.fit_transform(train_set.object_text.unique())\n",
    "hah.shape, hih.shape, huh.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ae9636f3-89ef-4c88-8974-79a1ae0596e7",
   "metadata": {
    "id": "W0DEbJkLg4Y6"
   },
   "outputs": [],
   "source": [
    "enc = OneHotEncoder(sparse_output=False)\n",
    "enc.fit(train_set[features_cat]);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a3950e1a-6646-43b6-a71a-ae37ebfbf306",
   "metadata": {
    "id": "W0DEbJkLg4Y6"
   },
   "outputs": [],
   "source": [
    "def get_features(dataset):\n",
    "    \"\"\" function to transform features for classifier\n",
    "    \"\"\"\n",
    "    dataset[\"rank_by_candidate_distance_filtered\"] = dataset.rank_by_candidate_distance_filtered.apply(lambda x: x if x != -1 else 100)#.value_counts()\n",
    "    X = dataset[features_int].astype(int)\n",
    "    X_cat = pd.DataFrame(enc.transform(dataset[features_cat]), columns = enc.get_feature_names_out(), index=X.index)\n",
    "    X = X.join(X_cat)\n",
    "    ## Text features\n",
    "    dataset[\"text_between_ents\"] = train_set[[\"sentence\", \"subject_begin\", \"subject_end\", \"object_begin\", \"object_end\"]].apply(text_between, axis=1)\n",
    "    X_text_between = vect.transform(dataset[\"text_between_ents\"])\n",
    "    X_text_between = pd.DataFrame(X_text_between.toarray(), columns=[f\"text_between_{w}\" for w in vect.get_feature_names_out()])\n",
    "    #X = X.join(X_text_between)\n",
    "    X_text_subject = vect_subj.transform(dataset[\"subject_text\"])\n",
    "    X_text_subject = pd.DataFrame(X_text_subject.toarray(), columns=[f\"text_subject_{w}\" for w in vect_subj.get_feature_names_out()])\n",
    "    X = X.join(X_text_subject)\n",
    "    X_text_object = vect_obj.transform(dataset[\"object_text\"])\n",
    "    X_text_object = pd.DataFrame(X_text_object.toarray(), columns=[f\"text_object_{w}\" for w in vect_obj.get_feature_names_out()])\n",
    "    X = X.join(X_text_object)\n",
    "    return X.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dca5994e-269c-4f29-a881-148ab6556da2",
   "metadata": {
    "id": "W0DEbJkLg4Y6",
    "scrolled": true
   },
   "source": [
    "### Correction method to force for specific labels: one prediction per subject\n",
    " * (All but PlugIn_of, Specification_of, Extension_of)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0b031d9e-fd8a-4744-ac66-d95272011b5f",
   "metadata": {
    "id": "W0DEbJkLg4Y6",
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "labels_with_one_relation = {\n",
    "    'Version_of',\n",
    "    'Developer_of',\n",
    "    'Citation_of',\n",
    "    'URL_of',\n",
    "     #'PlugIn_of',\n",
    "     #'Specification_of',\n",
    "     'Release_of',\n",
    "     #'Extension_of',\n",
    "     'License_of',\n",
    "     'Abbreviation_of',\n",
    "     'AlternativeName_of'}\n",
    "def force_one(pred):\n",
    "    if pred.iloc[0].target_label_binary in labels_with_one_relation:\n",
    "        pred = pred.sort_values(\"proba\", ascending=False)\n",
    "        pred[\"prediction_force_one\"] = 0\n",
    "        index_first = pred.index[0]\n",
    "        pred.loc[index_first, \"prediction_force_one\"] = 1\n",
    "    else:\n",
    "        pred[\"prediction_force_one\"] = (pred.proba >= 0.5).astype(int)\n",
    "    return pred.set_index(\"object_begin\").prediction_force_one\n",
    "\n",
    "def force_one_label(dataset, prediction, prediction_proba):\n",
    "    data = dataset.copy()\n",
    "    data[\"prediction\"] = prediction\n",
    "    data[\"proba\"] = prediction_proba[:,1]\n",
    "    subject_key = [\"sentence_id\", \"subject_begin\"]\n",
    "    pred_force_one = data.groupby(subject_key)[[\"target_label_binary\", \"object_begin\", \"proba\"]].apply(force_one)\n",
    "    data = data.set_index(subject_key + [\"object_begin\"]).join(pred_force_one).reset_index()\n",
    "    prediction_label_force_one = data[[\"target_label_binary\", \"prediction_force_one\"]]\\\n",
    "        .apply(lambda x: x.target_label_binary if x.prediction_force_one == 1 else 'nil', axis=1)\n",
    "    return data.prediction_force_one, prediction_label_force_one"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "729fe026-be79-465f-b2ae-7dab7aec68b4",
   "metadata": {
    "id": "W0DEbJkLg4Y6"
   },
   "source": [
    "## Prepare data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5b846d36-87cf-4e4a-aeb6-7b4c04334430",
   "metadata": {
    "id": "W0DEbJkLg4Y6"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1446, 960)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_train = get_features(train_set)\n",
    "y_train = train_set.label_binary\n",
    "\n",
    "X_valid = get_features(valid_set)\n",
    "y_valid_label_pred = valid_set.target_label_binary.to_numpy()\n",
    "y_valid = valid_set.label_binary\n",
    "y_valid_label = valid_set.relation.to_numpy()\n",
    "\n",
    "X_test = get_features(test_set)\n",
    "y_test_label_pred = test_set.target_label_binary.to_numpy()\n",
    "y_test = test_set.label_binary\n",
    "y_test_label = test_set.relation.to_numpy()\n",
    "\n",
    "len(y_valid_label), len(y_test_label)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f254bccb-ae46-42ae-beee-c04507854e4c",
   "metadata": {
    "id": "W0DEbJkLg4Y6"
   },
   "source": [
    "# Train Random Forest (use 5 runs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4fe49f63-a152-43f9-973e-199e5991bb1e",
   "metadata": {
    "id": "W0DEbJkLg4Y6"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: total: 30.9 s\n",
      "Wall time: 31.6 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "model = RandomForestClassifier(\n",
    "    #n_estimators=1000,\n",
    "    #max_depth=10,\n",
    "    #min_samples_split=10, #2\n",
    "    #min_samples_leaf=10, #1 \n",
    ")\n",
    "eval_results = []\n",
    "for run in range(5):\n",
    "    model.fit(X_train, y_train)\n",
    "    y_valid_pred = model.predict(X_valid)\n",
    "    y_valid_pred_label = [\"nil\" if p == 0 else y_valid_label_pred[idx]  for idx, p in enumerate(y_valid_pred)] \n",
    "    y_valid_pred_proba = model.predict_proba(X_valid)\n",
    "    y_valid_pred_force_one, y_valid_pred_label_force_one = force_one_label(valid_set, y_valid_pred, y_valid_pred_proba)\n",
    "    eval_result = pd.DataFrame(classification_report(y_valid_label, y_valid_pred_label_force_one, zero_division=0., labels=labels, output_dict=True)).T\n",
    "    eval_result = eval_result.reset_index().rename(columns=dict(index=\"label\"))\n",
    "    eval_results.append(eval_result) \n",
    "    y_test_pred = model.predict(X_test)\n",
    "    y_test_pred_proba = model.predict_proba(X_test)\n",
    "    y_test_pred_label = [\"nil\" if p == 0 else y_test_label_pred[idx]  for idx, p in enumerate(y_test_pred)]\n",
    "    y_test_pred_force_one, y_test_pred_label_force_one = force_one_label(test_set, y_test_pred, y_test_pred_proba)\n",
    "eval_results = pd.concat(eval_results)\n",
    "eval_results = eval_results.groupby(\"label\").describe()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e33adaf-3bcd-46f1-a32e-731070285319",
   "metadata": {
    "id": "W0DEbJkLg4Y6"
   },
   "source": [
    "# Evaluate\n",
    "## Validation Set (5 runs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "836ab95c-3ef9-44cb-8b20-4cd884178b6e",
   "metadata": {
    "id": "W0DEbJkLg4Y6"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">precision</th>\n",
       "      <th colspan=\"2\" halign=\"left\">recall</th>\n",
       "      <th colspan=\"2\" halign=\"left\">f1-score</th>\n",
       "      <th>support</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>mean</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Abbreviation_of</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AlternativeName_of</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Citation_of</th>\n",
       "      <td>0.933333</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.913043</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.923077</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>92.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Developer_of</th>\n",
       "      <td>0.964706</td>\n",
       "      <td>1.241267e-16</td>\n",
       "      <td>0.942529</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.953488</td>\n",
       "      <td>1.241267e-16</td>\n",
       "      <td>174.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Extension_of</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>License_of</th>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PlugIn_of</th>\n",
       "      <td>0.831378</td>\n",
       "      <td>1.491982e-02</td>\n",
       "      <td>0.950000</td>\n",
       "      <td>0.019562</td>\n",
       "      <td>0.886654</td>\n",
       "      <td>1.401586e-02</td>\n",
       "      <td>28.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Release_of</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Specification_of</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.488889</td>\n",
       "      <td>0.024845</td>\n",
       "      <td>0.656410</td>\n",
       "      <td>2.293403e-02</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>URL_of</th>\n",
       "      <td>0.918033</td>\n",
       "      <td>1.159191e-02</td>\n",
       "      <td>0.918033</td>\n",
       "      <td>0.011592</td>\n",
       "      <td>0.918033</td>\n",
       "      <td>1.159191e-02</td>\n",
       "      <td>61.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Version_of</th>\n",
       "      <td>0.991093</td>\n",
       "      <td>3.387288e-03</td>\n",
       "      <td>0.991093</td>\n",
       "      <td>0.003387</td>\n",
       "      <td>0.991093</td>\n",
       "      <td>3.387288e-03</td>\n",
       "      <td>247.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>macro avg</th>\n",
       "      <td>0.958049</td>\n",
       "      <td>8.074373e-04</td>\n",
       "      <td>0.918508</td>\n",
       "      <td>0.001813</td>\n",
       "      <td>0.929887</td>\n",
       "      <td>2.074251e-03</td>\n",
       "      <td>668.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>micro avg</th>\n",
       "      <td>0.961023</td>\n",
       "      <td>1.368375e-03</td>\n",
       "      <td>0.944910</td>\n",
       "      <td>0.001640</td>\n",
       "      <td>0.952898</td>\n",
       "      <td>1.478771e-03</td>\n",
       "      <td>668.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>weighted avg</th>\n",
       "      <td>0.962282</td>\n",
       "      <td>1.479255e-03</td>\n",
       "      <td>0.944910</td>\n",
       "      <td>0.001640</td>\n",
       "      <td>0.951006</td>\n",
       "      <td>1.412049e-03</td>\n",
       "      <td>668.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   precision                  recall            f1-score  \\\n",
       "                        mean           std      mean       std      mean   \n",
       "label                                                                      \n",
       "Abbreviation_of     1.000000  0.000000e+00  1.000000  0.000000  1.000000   \n",
       "AlternativeName_of  1.000000  0.000000e+00  1.000000  0.000000  1.000000   \n",
       "Citation_of         0.933333  0.000000e+00  0.913043  0.000000  0.923077   \n",
       "Developer_of        0.964706  1.241267e-16  0.942529  0.000000  0.953488   \n",
       "Extension_of        1.000000  0.000000e+00  1.000000  0.000000  1.000000   \n",
       "License_of          0.900000  0.000000e+00  0.900000  0.000000  0.900000   \n",
       "PlugIn_of           0.831378  1.491982e-02  0.950000  0.019562  0.886654   \n",
       "Release_of          1.000000  0.000000e+00  1.000000  0.000000  1.000000   \n",
       "Specification_of    1.000000  0.000000e+00  0.488889  0.024845  0.656410   \n",
       "URL_of              0.918033  1.159191e-02  0.918033  0.011592  0.918033   \n",
       "Version_of          0.991093  3.387288e-03  0.991093  0.003387  0.991093   \n",
       "macro avg           0.958049  8.074373e-04  0.918508  0.001813  0.929887   \n",
       "micro avg           0.961023  1.368375e-03  0.944910  0.001640  0.952898   \n",
       "weighted avg        0.962282  1.479255e-03  0.944910  0.001640  0.951006   \n",
       "\n",
       "                                 support  \n",
       "                             std    mean  \n",
       "label                                     \n",
       "Abbreviation_of     0.000000e+00     6.0  \n",
       "AlternativeName_of  0.000000e+00     3.0  \n",
       "Citation_of         0.000000e+00    92.0  \n",
       "Developer_of        1.241267e-16   174.0  \n",
       "Extension_of        0.000000e+00    13.0  \n",
       "License_of          0.000000e+00    10.0  \n",
       "PlugIn_of           1.401586e-02    28.0  \n",
       "Release_of          0.000000e+00    16.0  \n",
       "Specification_of    2.293403e-02    18.0  \n",
       "URL_of              1.159191e-02    61.0  \n",
       "Version_of          3.387288e-03   247.0  \n",
       "macro avg           2.074251e-03   668.0  \n",
       "micro avg           1.478771e-03   668.0  \n",
       "weighted avg        1.412049e-03   668.0  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "show = [\n",
    "    (\"precision\", \"mean\"),\n",
    "    (\"precision\", \"std\"),\n",
    "    (\"recall\", \"mean\"),\n",
    "    (\"recall\", \"std\"),\n",
    "    (\"f1-score\", \"mean\"),\n",
    "    (\"f1-score\", \"std\"),\n",
    "    (\"support\", \"mean\"),\n",
    "]\n",
    "eval_results[show]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8613027-ed58-4079-8bc1-f24b093a6f30",
   "metadata": {
    "id": "W0DEbJkLg4Y6",
    "scrolled": true
   },
   "source": [
    "## Test Set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "861c2602-03dd-4cf3-95d6-5498c3cd19a3",
   "metadata": {
    "id": "W0DEbJkLg4Y6"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1-score</th>\n",
       "      <th>support</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Version_of</th>\n",
       "      <td>0.981651</td>\n",
       "      <td>0.981651</td>\n",
       "      <td>0.981651</td>\n",
       "      <td>109.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Developer_of</th>\n",
       "      <td>0.939759</td>\n",
       "      <td>0.939759</td>\n",
       "      <td>0.939759</td>\n",
       "      <td>83.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Citation_of</th>\n",
       "      <td>0.977778</td>\n",
       "      <td>0.977778</td>\n",
       "      <td>0.977778</td>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>URL_of</th>\n",
       "      <td>0.897436</td>\n",
       "      <td>0.897436</td>\n",
       "      <td>0.897436</td>\n",
       "      <td>39.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PlugIn_of</th>\n",
       "      <td>0.705882</td>\n",
       "      <td>0.923077</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Specification_of</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>0.545455</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Release_of</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Extension_of</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>0.952381</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>License_of</th>\n",
       "      <td>0.727273</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Abbreviation_of</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AlternativeName_of</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>micro avg</th>\n",
       "      <td>0.939394</td>\n",
       "      <td>0.939394</td>\n",
       "      <td>0.939394</td>\n",
       "      <td>330.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>macro avg</th>\n",
       "      <td>0.929980</td>\n",
       "      <td>0.899335</td>\n",
       "      <td>0.899496</td>\n",
       "      <td>330.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>weighted avg</th>\n",
       "      <td>0.944612</td>\n",
       "      <td>0.939394</td>\n",
       "      <td>0.937696</td>\n",
       "      <td>330.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    precision    recall  f1-score  support\n",
       "Version_of           0.981651  0.981651  0.981651    109.0\n",
       "Developer_of         0.939759  0.939759  0.939759     83.0\n",
       "Citation_of          0.977778  0.977778  0.977778     45.0\n",
       "URL_of               0.897436  0.897436  0.897436     39.0\n",
       "PlugIn_of            0.705882  0.923077  0.800000     13.0\n",
       "Specification_of     1.000000  0.375000  0.545455      8.0\n",
       "Release_of           1.000000  1.000000  1.000000      7.0\n",
       "Extension_of         1.000000  0.909091  0.952381     11.0\n",
       "License_of           0.727273  0.888889  0.800000      9.0\n",
       "Abbreviation_of      1.000000  1.000000  1.000000      5.0\n",
       "AlternativeName_of   1.000000  1.000000  1.000000      1.0\n",
       "micro avg            0.939394  0.939394  0.939394    330.0\n",
       "macro avg            0.929980  0.899335  0.899496    330.0\n",
       "weighted avg         0.944612  0.939394  0.937696    330.0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(classification_report(y_test_label, y_test_pred_label_force_one, zero_division=0., labels=labels, output_dict=True)).T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37d91106-6a9a-4134-97c3-b8c8361d9f1e",
   "metadata": {
    "id": "W0DEbJkLg4Y6",
    "scrolled": true
   },
   "source": [
    "### Best performance: 0.927"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f73db204-6997-4f69-96af-fb79df46993d",
   "metadata": {
    "id": "W0DEbJkLg4Y6"
   },
   "source": [
    "## Look into wrong predictions (on test set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1053e652-64db-4a45-b6fe-d550a4f8ce16",
   "metadata": {
    "id": "W0DEbJkLg4Y6"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "61"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ha = valid_set.copy()\n",
    "ha[\"pred\"] = y_valid_pred_label\n",
    "cols = [\"sentence\", \"subject_text\", \"target_label_binary\", \"object_text\", \"relation\", \"pred\"]\n",
    "ha = ha[list(y_valid_label != y_valid_pred_label_force_one)][cols]#.sample(3)\n",
    "len(ha)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ea9dc02a-dfd7-4232-b1c7-95f8009c0018",
   "metadata": {
    "id": "W0DEbJkLg4Y6"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "target_label_binary\n",
       "Developer_of        16\n",
       "Citation_of         14\n",
       "Specification_of    10\n",
       "URL_of              10\n",
       "PlugIn_of            7\n",
       "License_of           2\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ha.target_label_binary.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0f0c71ba-a5db-4a39-a8d4-db8f7bc19d5b",
   "metadata": {
    "id": "W0DEbJkLg4Y6",
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#pd.Series(y_test).value_counts(), pd.Series(y_test_pred).value_counts(), y_test_pred_force_one.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1236a4fe-811b-4e7a-8c46-cab18d36ceac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sentence</th>\n",
       "      <th>subject_text</th>\n",
       "      <th>target_label_binary</th>\n",
       "      <th>object_text</th>\n",
       "      <th>relation</th>\n",
       "      <th>pred</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PyPhi is open - source and licensed under the GPLv 3 ; the source code is hosted on GitHub at https://github.com/wmayner/pyphi .</td>\n",
       "      <td>PyPhi</td>\n",
       "      <td>PlugIn_of</td>\n",
       "      <td>GitHub</td>\n",
       "      <td>nil</td>\n",
       "      <td>PlugIn_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CoXpress is written in the native R language and has been fully tested on both windows and linux .</td>\n",
       "      <td>CoXpress</td>\n",
       "      <td>PlugIn_of</td>\n",
       "      <td>R</td>\n",
       "      <td>nil</td>\n",
       "      <td>PlugIn_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Mindboggle is a freely downloadable , open source software package written in Matlab ( version 6 , release 13 , with the Image Processing Toolbox , The Mathworks Inc . , USA ) and has been tested on different models of desktop and laptop computers running different distributions of Linux , as we...</td>\n",
       "      <td>The Mathworks Inc .</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>Matlab</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Mindboggle is a freely downloadable , open source software package written in Matlab ( version 6 , release 13 , with the Image Processing Toolbox , The Mathworks Inc . , USA ) and has been tested on different models of desktop and laptop computers running different distributions of Linux , as we...</td>\n",
       "      <td>The Mathworks Inc .</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>Image Processing Toolbox</td>\n",
       "      <td>nil</td>\n",
       "      <td>Developer_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>Availability : FunciSNP is available from Bioconductor ( bioconductor.org ) .</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>PlugIn_of</td>\n",
       "      <td>FunciSNP</td>\n",
       "      <td>nil</td>\n",
       "      <td>PlugIn_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>More details : Software name : CBFA plugin for Optflux Project home page : http://www.optflux.org/ Methods details and application tutorial : http://www.optflux.org/cbfa Operating system ( s ) : Platform independent Programming languages : Java Other requirements : Java JRE 1.7 .x ( for Mac OS u...</td>\n",
       "      <td>CBFA</td>\n",
       "      <td>PlugIn_of</td>\n",
       "      <td>Optflux</td>\n",
       "      <td>nil</td>\n",
       "      <td>PlugIn_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>More details : Software name : CBFA plugin for Optflux Project home page : http://www.optflux.org/ Methods details and application tutorial : http://www.optflux.org/cbfa Operating system ( s ) : Platform independent Programming languages : Java Other requirements : Java JRE 1.7 .x ( for Mac OS u...</td>\n",
       "      <td>http://www.optflux.org/cbfa</td>\n",
       "      <td>URL_of</td>\n",
       "      <td>CBFA</td>\n",
       "      <td>URL_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>More details : Software name : CBFA plugin for Optflux Project home page : http://www.optflux.org/ Methods details and application tutorial : http://www.optflux.org/cbfa Operating system ( s ) : Platform independent Programming languages : Java Other requirements : Java JRE 1.7 .x ( for Mac OS u...</td>\n",
       "      <td>http://www.optflux.org/cbfa</td>\n",
       "      <td>URL_of</td>\n",
       "      <td>Optflux</td>\n",
       "      <td>nil</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>\" • Project name : SSPACE - LongRead • Project home page : http://www.baseclear.com/bioinformatics-tools/ • Operating systems : All major Linux platforms • Programming languages : Perl , C + + ( the latter is required for BLASR , see below ) • Other requirements : BLASR for the alignment of long...</td>\n",
       "      <td>BaseTools</td>\n",
       "      <td>License_of</td>\n",
       "      <td>SSPACE - LongRead</td>\n",
       "      <td>License_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>\" • Project name : SSPACE - LongRead • Project home page : http://www.baseclear.com/bioinformatics-tools/ • Operating systems : All major Linux platforms • Programming languages : Perl , C + + ( the latter is required for BLASR , see below ) • Other requirements : BLASR for the alignment of long...</td>\n",
       "      <td>BaseTools</td>\n",
       "      <td>License_of</td>\n",
       "      <td>BLASR</td>\n",
       "      <td>nil</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>Here , we present a novel Python - based toolbox called HDDM ( hierarchical drift diffusion model ) , which allows fast and flexible estimation of the the drift - diffusion model and the related linear ballistic accumulator model .</td>\n",
       "      <td>HDDM</td>\n",
       "      <td>PlugIn_of</td>\n",
       "      <td>Python</td>\n",
       "      <td>nil</td>\n",
       "      <td>PlugIn_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>Expression Analysis of mRNAs and miRNAs : Statistical analyses to compare mRNA and miRNA expression profiles were performed using the \" R \" statistical environment edgeR [96] , vegan [97] and gplots [98] packages .</td>\n",
       "      <td>gplots</td>\n",
       "      <td>PlugIn_of</td>\n",
       "      <td>R</td>\n",
       "      <td>PlugIn_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428</th>\n",
       "      <td>All data processing and statistical analyses were performed with the Statistical Package for Social Science ( SPSS ) software , vers . 18.0 ( SPSS , Chicago , IL , USA ) and SAS vers . 8.2 ( SAS System for Windows , SAS Institute , Cary , NC , USA ) .</td>\n",
       "      <td>SPSS</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>Statistical Package for Social Science</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442</th>\n",
       "      <td>All data processing and statistical analyses were performed with the Statistical Package for Social Science ( SPSS ) software , vers . 18.0 ( SPSS , Chicago , IL , USA ) and SAS vers . 8.2 ( SAS System for Windows , SAS Institute , Cary , NC , USA ) .</td>\n",
       "      <td>SPSS</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>SAS</td>\n",
       "      <td>nil</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447</th>\n",
       "      <td>All data processing and statistical analyses were performed with the Statistical Package for Social Science ( SPSS ) software , vers . 18.0 ( SPSS , Chicago , IL , USA ) and SAS vers . 8.2 ( SAS System for Windows , SAS Institute , Cary , NC , USA ) .</td>\n",
       "      <td>SAS</td>\n",
       "      <td>Specification_of</td>\n",
       "      <td>SAS</td>\n",
       "      <td>Specification_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>473</th>\n",
       "      <td>SMI software ( Experiment Center ™ and iView X ™ ) were used to collect and record calibration , present the stimuli , and record gaze data .</td>\n",
       "      <td>SMI</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>iView X</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>523</th>\n",
       "      <td>The vocalization synthesis relied on the articulatory synthesizer developed by Boersma and available in Praat [ 61 , 66 ] .</td>\n",
       "      <td>[ 61 , 66 ]</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>articulatory synthesizer</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>578</th>\n",
       "      <td>As a measure of the test - retest reliability of the degree of lateralization , we computed an ICC ( two - way mixed model with absolute agreement using SPSS ; IBM SPSS Statistics for Macintosh , version 22.0 ) for the LIs in the frontal and parietal ROI , respectively .</td>\n",
       "      <td>SPSS Statistics</td>\n",
       "      <td>Specification_of</td>\n",
       "      <td>SPSS</td>\n",
       "      <td>Specification_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>614</th>\n",
       "      <td>Figure generated using clam for R [53] .</td>\n",
       "      <td>[53]</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>clam</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>615</th>\n",
       "      <td>Figure generated using clam for R [53] .</td>\n",
       "      <td>[53]</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>R</td>\n",
       "      <td>nil</td>\n",
       "      <td>Citation_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>617</th>\n",
       "      <td>Cluster analysis was used to determine boundaries between diatom assemblage zones in the stratigraphy , with the number of significant zones evaluated by the broken stick model , using the rioja package in R [ 62 , 63 ] .</td>\n",
       "      <td>[ 62 , 63 ]</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>rioja</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>624</th>\n",
       "      <td>To calculate the ARI scores , we used the function “ adjustedRandIndex ” in R software [58] , [59] and , for the ICC and AIC , we use the R function “ lmer ” [60] that employs a Restricted Maximum Likelihood procedure [61] to obtain estimates of , and AIC .</td>\n",
       "      <td>[58]</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>adjustedRandIndex</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>625</th>\n",
       "      <td>To calculate the ARI scores , we used the function “ adjustedRandIndex ” in R software [58] , [59] and , for the ICC and AIC , we use the R function “ lmer ” [60] that employs a Restricted Maximum Likelihood procedure [61] to obtain estimates of , and AIC .</td>\n",
       "      <td>[59]</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>adjustedRandIndex</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630</th>\n",
       "      <td>To calculate the ARI scores , we used the function “ adjustedRandIndex ” in R software [58] , [59] and , for the ICC and AIC , we use the R function “ lmer ” [60] that employs a Restricted Maximum Likelihood procedure [61] to obtain estimates of , and AIC .</td>\n",
       "      <td>[58]</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>R</td>\n",
       "      <td>nil</td>\n",
       "      <td>Citation_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>631</th>\n",
       "      <td>To calculate the ARI scores , we used the function “ adjustedRandIndex ” in R software [58] , [59] and , for the ICC and AIC , we use the R function “ lmer ” [60] that employs a Restricted Maximum Likelihood procedure [61] to obtain estimates of , and AIC .</td>\n",
       "      <td>[59]</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>R</td>\n",
       "      <td>nil</td>\n",
       "      <td>Citation_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>671</th>\n",
       "      <td>All statistical tests , carried out in R Studio Version 0.98.1102 and SPSS ( IBM SPSS Statistics for Macintosh , version 22.0 ) , were two - tailed and the statistical significance level was set at 0.05 .</td>\n",
       "      <td>SPSS Statistics</td>\n",
       "      <td>Specification_of</td>\n",
       "      <td>SPSS</td>\n",
       "      <td>Specification_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>The software used for spectrum visualization and MS / MSMS protein identification was Flex Analysis ( v. 3.3 ) and BioTools ( Bruker Daltonics ) linked to MASCOT ( Matrix Science , Boston , MA 2016 ( http://www.matrixscience.com/ ) to search against the NCBInr protein sequence databases .</td>\n",
       "      <td>http://www.matrixscience.com/</td>\n",
       "      <td>URL_of</td>\n",
       "      <td>MASCOT</td>\n",
       "      <td>URL_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>720</th>\n",
       "      <td>The software used for spectrum visualization and MS / MSMS protein identification was Flex Analysis ( v. 3.3 ) and BioTools ( Bruker Daltonics ) linked to MASCOT ( Matrix Science , Boston , MA 2016 ( http://www.matrixscience.com/ ) to search against the NCBInr protein sequence databases .</td>\n",
       "      <td>http://www.matrixscience.com/</td>\n",
       "      <td>URL_of</td>\n",
       "      <td>Matrix Science</td>\n",
       "      <td>nil</td>\n",
       "      <td>URL_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>723</th>\n",
       "      <td>All statistical analysis was performed using R ver. 2.15.1 ( R Development Core Team ; www.r-project.org ) with critical P - values set to 0.05 , unless otherwise stated .</td>\n",
       "      <td>www.r-project.org</td>\n",
       "      <td>URL_of</td>\n",
       "      <td>R</td>\n",
       "      <td>URL_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>724</th>\n",
       "      <td>All statistical analysis was performed using R ver. 2.15.1 ( R Development Core Team ; www.r-project.org ) with critical P - values set to 0.05 , unless otherwise stated .</td>\n",
       "      <td>www.r-project.org</td>\n",
       "      <td>URL_of</td>\n",
       "      <td>R Development Core Team</td>\n",
       "      <td>nil</td>\n",
       "      <td>URL_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>730</th>\n",
       "      <td>To extract noun phrases to use as search terms for each of the papers on PubMed Central , we run a software program written in Python based NLTK ( http://www.nltk.org/ ) to automatically extract nouns , adjectives , and noun phrases from each sentence in the paper .</td>\n",
       "      <td>http://www.nltk.org/</td>\n",
       "      <td>URL_of</td>\n",
       "      <td>Python</td>\n",
       "      <td>URL_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731</th>\n",
       "      <td>To extract noun phrases to use as search terms for each of the papers on PubMed Central , we run a software program written in Python based NLTK ( http://www.nltk.org/ ) to automatically extract nouns , adjectives , and noun phrases from each sentence in the paper .</td>\n",
       "      <td>http://www.nltk.org/</td>\n",
       "      <td>URL_of</td>\n",
       "      <td>NLTK</td>\n",
       "      <td>nil</td>\n",
       "      <td>URL_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>Then , the hypotheses were tested with OLS Ordinary Least Square through the PROCESS macro for SPSS , which assesses the conditional indirect effects [38] .</td>\n",
       "      <td>[38]</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>PROCESS</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>Then , the hypotheses were tested with OLS Ordinary Least Square through the PROCESS macro for SPSS , which assesses the conditional indirect effects [38] .</td>\n",
       "      <td>[38]</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>SPSS</td>\n",
       "      <td>nil</td>\n",
       "      <td>Citation_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>867</th>\n",
       "      <td>Analysis was then done by one researcher ( SS ) using MSOffice ( Word and Excel , Microsoft Corp . ) and NVivo ( QSR International ) software for qualitative data analysis .</td>\n",
       "      <td>Microsoft Corp .</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>MSOffice</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>873</th>\n",
       "      <td>Analysis was then done by one researcher ( SS ) using MSOffice ( Word and Excel , Microsoft Corp . ) and NVivo ( QSR International ) software for qualitative data analysis .</td>\n",
       "      <td>Microsoft Corp .</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>Word</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>877</th>\n",
       "      <td>Analysis was then done by one researcher ( SS ) using MSOffice ( Word and Excel , Microsoft Corp . ) and NVivo ( QSR International ) software for qualitative data analysis .</td>\n",
       "      <td>Microsoft Corp .</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>Excel</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>881</th>\n",
       "      <td>Analysis was then done by one researcher ( SS ) using MSOffice ( Word and Excel , Microsoft Corp . ) and NVivo ( QSR International ) software for qualitative data analysis .</td>\n",
       "      <td>Microsoft Corp .</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>NVivo</td>\n",
       "      <td>nil</td>\n",
       "      <td>Developer_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>955</th>\n",
       "      <td>The data were analyzed using SPSS software ( Statistical Package for the Social Sciences , version 17.0 ) .</td>\n",
       "      <td>Statistical Package for the Social Sciences</td>\n",
       "      <td>Specification_of</td>\n",
       "      <td>SPSS</td>\n",
       "      <td>Specification_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>976</th>\n",
       "      <td>All three stages of analysis were conducted using SPSS software ( IBM SPSS Statistics 20 , IBM Corporation ) .</td>\n",
       "      <td>SPSS Statistics</td>\n",
       "      <td>Specification_of</td>\n",
       "      <td>SPSS</td>\n",
       "      <td>Specification_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1069</th>\n",
       "      <td>The images were subsequently processed and analyzed on a pixel - by - pixel basis using the Matlab Image Processing Toolbox ( Mathworks , Matlab v 2011b ) using script modified from Theyel and colleagues [20] .</td>\n",
       "      <td>Mathworks</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>Matlab</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1071</th>\n",
       "      <td>The images were subsequently processed and analyzed on a pixel - by - pixel basis using the Matlab Image Processing Toolbox ( Mathworks , Matlab v 2011b ) using script modified from Theyel and colleagues [20] .</td>\n",
       "      <td>Matlab</td>\n",
       "      <td>Specification_of</td>\n",
       "      <td>Matlab</td>\n",
       "      <td>Specification_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1073</th>\n",
       "      <td>The images were subsequently processed and analyzed on a pixel - by - pixel basis using the Matlab Image Processing Toolbox ( Mathworks , Matlab v 2011b ) using script modified from Theyel and colleagues [20] .</td>\n",
       "      <td>Mathworks</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>Image Processing Toolbox</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1076</th>\n",
       "      <td>The images were subsequently processed and analyzed on a pixel - by - pixel basis using the Matlab Image Processing Toolbox ( Mathworks , Matlab v 2011b ) using script modified from Theyel and colleagues [20] .</td>\n",
       "      <td>Mathworks</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>Matlab</td>\n",
       "      <td>nil</td>\n",
       "      <td>Developer_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1139</th>\n",
       "      <td>Statistical analyses were conducted using Statistical Package for the Social Sciences Version 20.0 ( SPSS 20.0 ) and Analysis of Moment Structures Version 20.0 ( AMOS 20.0 ) .</td>\n",
       "      <td>AMOS</td>\n",
       "      <td>Specification_of</td>\n",
       "      <td>Analysis of Moment Structures</td>\n",
       "      <td>Specification_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1154</th>\n",
       "      <td>For the pre - processing and statistical analyses , the statistical parametric mapping software package ( SPM 5 , Wellcome Department of Cognitive Neurology , London ; www.fil.ion.ucl.ac.uk/spm ) was used and implemented in Matlab ( Mathworks , Inc . , Natick , MA , USA release 14 ) .</td>\n",
       "      <td>SPM</td>\n",
       "      <td>Specification_of</td>\n",
       "      <td>statistical parametric mapping</td>\n",
       "      <td>Specification_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1174</th>\n",
       "      <td>All analyses were performed using SAS / STAT ™ software version 9.1 ( SAS Institute Inc . , Cary , NC ) , Statistical Package for the Social Sciences ™ for Windows ™ ( SPSS ™ version 14.1 , SPSS Inc . , Chicago , IL ) , and RevMan version 4.1 ( Cochrane Collaboration , Oxford , UK ) .</td>\n",
       "      <td>SAS Institute Inc .</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>SAS</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1193</th>\n",
       "      <td>All analyses were performed using SAS / STAT ™ software version 9.1 ( SAS Institute Inc . , Cary , NC ) , Statistical Package for the Social Sciences ™ for Windows ™ ( SPSS ™ version 14.1 , SPSS Inc . , Chicago , IL ) , and RevMan version 4.1 ( Cochrane Collaboration , Oxford , UK ) .</td>\n",
       "      <td>SAS Institute Inc .</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>Statistical Package for the Social Sciences</td>\n",
       "      <td>nil</td>\n",
       "      <td>Developer_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1228</th>\n",
       "      <td>DLMs were implemented using the dlnm package version 2.2.6 in R Studio ( Vienna , Austria ) and other analyses were performed in SAS ( v 9.4 , SAS Institute Inc . , Cary , NC ) .</td>\n",
       "      <td>dlnm</td>\n",
       "      <td>PlugIn_of</td>\n",
       "      <td>R Studio</td>\n",
       "      <td>nil</td>\n",
       "      <td>PlugIn_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1345</th>\n",
       "      <td>The IVDDM and IVDGE of feedstuffs were compared using the MIXED procedure of SAS ( SAS v 9.2 , SAS Institute Inc . , Cary , NC , USA ) , where feedstuffs were treated as fixed factor and batch as random factor .</td>\n",
       "      <td>SAS</td>\n",
       "      <td>Specification_of</td>\n",
       "      <td>SAS</td>\n",
       "      <td>Specification_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1351</th>\n",
       "      <td>We generated multilevel logistic regression models with random intercepts to evaluate the impact of patients and trust factors on each dependent variable using MLWin V. 2.31 software ( Rasbash J , Charlton C , Browne W , et al . MLwiN , Version 2.31 . Centre for Multilevel Modelling , University...</td>\n",
       "      <td>Rasbash J , Charlton C , Browne W , et al .</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>MLWin</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1353</th>\n",
       "      <td>We generated multilevel logistic regression models with random intercepts to evaluate the impact of patients and trust factors on each dependent variable using MLWin V. 2.31 software ( Rasbash J , Charlton C , Browne W , et al . MLwiN , Version 2.31 . Centre for Multilevel Modelling , University...</td>\n",
       "      <td>MLwiN</td>\n",
       "      <td>Specification_of</td>\n",
       "      <td>MLWin</td>\n",
       "      <td>Specification_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1357</th>\n",
       "      <td>We generated multilevel logistic regression models with random intercepts to evaluate the impact of patients and trust factors on each dependent variable using MLWin V. 2.31 software ( Rasbash J , Charlton C , Browne W , et al . MLwiN , Version 2.31 . Centre for Multilevel Modelling , University...</td>\n",
       "      <td>Rasbash J , Charlton C , Browne W , et al .</td>\n",
       "      <td>Developer_of</td>\n",
       "      <td>MLwiN</td>\n",
       "      <td>nil</td>\n",
       "      <td>Developer_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1399</th>\n",
       "      <td>CIBERSORT analyses were performed on RNAseq data by using the analytical tool developed by Newman et al . ( http://cibersort.stanford.edu ) [30] .</td>\n",
       "      <td>http://cibersort.stanford.edu</td>\n",
       "      <td>URL_of</td>\n",
       "      <td>CIBERSORT</td>\n",
       "      <td>URL_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1401</th>\n",
       "      <td>CIBERSORT analyses were performed on RNAseq data by using the analytical tool developed by Newman et al . ( http://cibersort.stanford.edu ) [30] .</td>\n",
       "      <td>http://cibersort.stanford.edu</td>\n",
       "      <td>URL_of</td>\n",
       "      <td>Newman et al .</td>\n",
       "      <td>nil</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1421</th>\n",
       "      <td>Data were imported and normalized using the ChAMP R package [59] .</td>\n",
       "      <td>[59]</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>ChAMP</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1422</th>\n",
       "      <td>Data were imported and normalized using the ChAMP R package [59] .</td>\n",
       "      <td>[59]</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>R</td>\n",
       "      <td>nil</td>\n",
       "      <td>Citation_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1438</th>\n",
       "      <td>Differential expression genes between the infected and control groups were identified with the DESeq of R package [13] .</td>\n",
       "      <td>[13]</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>DESeq</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>nil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1439</th>\n",
       "      <td>Differential expression genes between the infected and control groups were identified with the DESeq of R package [13] .</td>\n",
       "      <td>[13]</td>\n",
       "      <td>Citation_of</td>\n",
       "      <td>R</td>\n",
       "      <td>nil</td>\n",
       "      <td>Citation_of</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                                                                                                                                                         sentence  \\\n",
       "9                                                                                                                                                                                PyPhi is open - source and licensed under the GPLv 3 ; the source code is hosted on GitHub at https://github.com/wmayner/pyphi .   \n",
       "16                                                                                                                                                                                                             CoXpress is written in the native R language and has been fully tested on both windows and linux .   \n",
       "30    Mindboggle is a freely downloadable , open source software package written in Matlab ( version 6 , release 13 , with the Image Processing Toolbox , The Mathworks Inc . , USA ) and has been tested on different models of desktop and laptop computers running different distributions of Linux , as we...   \n",
       "39    Mindboggle is a freely downloadable , open source software package written in Matlab ( version 6 , release 13 , with the Image Processing Toolbox , The Mathworks Inc . , USA ) and has been tested on different models of desktop and laptop computers running different distributions of Linux , as we...   \n",
       "54                                                                                                                                                                                                                                  Availability : FunciSNP is available from Bioconductor ( bioconductor.org ) .   \n",
       "94    More details : Software name : CBFA plugin for Optflux Project home page : http://www.optflux.org/ Methods details and application tutorial : http://www.optflux.org/cbfa Operating system ( s ) : Platform independent Programming languages : Java Other requirements : Java JRE 1.7 .x ( for Mac OS u...   \n",
       "96    More details : Software name : CBFA plugin for Optflux Project home page : http://www.optflux.org/ Methods details and application tutorial : http://www.optflux.org/cbfa Operating system ( s ) : Platform independent Programming languages : Java Other requirements : Java JRE 1.7 .x ( for Mac OS u...   \n",
       "106   More details : Software name : CBFA plugin for Optflux Project home page : http://www.optflux.org/ Methods details and application tutorial : http://www.optflux.org/cbfa Operating system ( s ) : Platform independent Programming languages : Java Other requirements : Java JRE 1.7 .x ( for Mac OS u...   \n",
       "157   \" • Project name : SSPACE - LongRead • Project home page : http://www.baseclear.com/bioinformatics-tools/ • Operating systems : All major Linux platforms • Programming languages : Perl , C + + ( the latter is required for BLASR , see below ) • Other requirements : BLASR for the alignment of long...   \n",
       "172   \" • Project name : SSPACE - LongRead • Project home page : http://www.baseclear.com/bioinformatics-tools/ • Operating systems : All major Linux platforms • Programming languages : Perl , C + + ( the latter is required for BLASR , see below ) • Other requirements : BLASR for the alignment of long...   \n",
       "187                                                                       Here , we present a novel Python - based toolbox called HDDM ( hierarchical drift diffusion model ) , which allows fast and flexible estimation of the the drift - diffusion model and the related linear ballistic accumulator model .   \n",
       "351                                                                                        Expression Analysis of mRNAs and miRNAs : Statistical analyses to compare mRNA and miRNA expression profiles were performed using the \" R \" statistical environment edgeR [96] , vegan [97] and gplots [98] packages .   \n",
       "428                                                   All data processing and statistical analyses were performed with the Statistical Package for Social Science ( SPSS ) software , vers . 18.0 ( SPSS , Chicago , IL , USA ) and SAS vers . 8.2 ( SAS System for Windows , SAS Institute , Cary , NC , USA ) .   \n",
       "442                                                   All data processing and statistical analyses were performed with the Statistical Package for Social Science ( SPSS ) software , vers . 18.0 ( SPSS , Chicago , IL , USA ) and SAS vers . 8.2 ( SAS System for Windows , SAS Institute , Cary , NC , USA ) .   \n",
       "447                                                   All data processing and statistical analyses were performed with the Statistical Package for Social Science ( SPSS ) software , vers . 18.0 ( SPSS , Chicago , IL , USA ) and SAS vers . 8.2 ( SAS System for Windows , SAS Institute , Cary , NC , USA ) .   \n",
       "473                                                                                                                                                                 SMI software ( Experiment Center ™ and iView X ™ ) were used to collect and record calibration , present the stimuli , and record gaze data .   \n",
       "523                                                                                                                                                                                   The vocalization synthesis relied on the articulatory synthesizer developed by Boersma and available in Praat [ 61 , 66 ] .   \n",
       "578                               As a measure of the test - retest reliability of the degree of lateralization , we computed an ICC ( two - way mixed model with absolute agreement using SPSS ; IBM SPSS Statistics for Macintosh , version 22.0 ) for the LIs in the frontal and parietal ROI , respectively .   \n",
       "614                                                                                                                                                                                                                                                                      Figure generated using clam for R [53] .   \n",
       "615                                                                                                                                                                                                                                                                      Figure generated using clam for R [53] .   \n",
       "617                                                                                 Cluster analysis was used to determine boundaries between diatom assemblage zones in the stratigraphy , with the number of significant zones evaluated by the broken stick model , using the rioja package in R [ 62 , 63 ] .   \n",
       "624                                             To calculate the ARI scores , we used the function “ adjustedRandIndex ” in R software [58] , [59] and , for the ICC and AIC , we use the R function “ lmer ” [60] that employs a Restricted Maximum Likelihood procedure [61] to obtain estimates of , and AIC .   \n",
       "625                                             To calculate the ARI scores , we used the function “ adjustedRandIndex ” in R software [58] , [59] and , for the ICC and AIC , we use the R function “ lmer ” [60] that employs a Restricted Maximum Likelihood procedure [61] to obtain estimates of , and AIC .   \n",
       "630                                             To calculate the ARI scores , we used the function “ adjustedRandIndex ” in R software [58] , [59] and , for the ICC and AIC , we use the R function “ lmer ” [60] that employs a Restricted Maximum Likelihood procedure [61] to obtain estimates of , and AIC .   \n",
       "631                                             To calculate the ARI scores , we used the function “ adjustedRandIndex ” in R software [58] , [59] and , for the ICC and AIC , we use the R function “ lmer ” [60] that employs a Restricted Maximum Likelihood procedure [61] to obtain estimates of , and AIC .   \n",
       "671                                                                                                  All statistical tests , carried out in R Studio Version 0.98.1102 and SPSS ( IBM SPSS Statistics for Macintosh , version 22.0 ) , were two - tailed and the statistical significance level was set at 0.05 .   \n",
       "719             The software used for spectrum visualization and MS / MSMS protein identification was Flex Analysis ( v. 3.3 ) and BioTools ( Bruker Daltonics ) linked to MASCOT ( Matrix Science , Boston , MA 2016 ( http://www.matrixscience.com/ ) to search against the NCBInr protein sequence databases .   \n",
       "720             The software used for spectrum visualization and MS / MSMS protein identification was Flex Analysis ( v. 3.3 ) and BioTools ( Bruker Daltonics ) linked to MASCOT ( Matrix Science , Boston , MA 2016 ( http://www.matrixscience.com/ ) to search against the NCBInr protein sequence databases .   \n",
       "723                                                                                                                                   All statistical analysis was performed using R ver. 2.15.1 ( R Development Core Team ; www.r-project.org ) with critical P - values set to 0.05 , unless otherwise stated .   \n",
       "724                                                                                                                                   All statistical analysis was performed using R ver. 2.15.1 ( R Development Core Team ; www.r-project.org ) with critical P - values set to 0.05 , unless otherwise stated .   \n",
       "730                                    To extract noun phrases to use as search terms for each of the papers on PubMed Central , we run a software program written in Python based NLTK ( http://www.nltk.org/ ) to automatically extract nouns , adjectives , and noun phrases from each sentence in the paper .   \n",
       "731                                    To extract noun phrases to use as search terms for each of the papers on PubMed Central , we run a software program written in Python based NLTK ( http://www.nltk.org/ ) to automatically extract nouns , adjectives , and noun phrases from each sentence in the paper .   \n",
       "861                                                                                                                                                  Then , the hypotheses were tested with OLS Ordinary Least Square through the PROCESS macro for SPSS , which assesses the conditional indirect effects [38] .   \n",
       "862                                                                                                                                                  Then , the hypotheses were tested with OLS Ordinary Least Square through the PROCESS macro for SPSS , which assesses the conditional indirect effects [38] .   \n",
       "867                                                                                                                                 Analysis was then done by one researcher ( SS ) using MSOffice ( Word and Excel , Microsoft Corp . ) and NVivo ( QSR International ) software for qualitative data analysis .   \n",
       "873                                                                                                                                 Analysis was then done by one researcher ( SS ) using MSOffice ( Word and Excel , Microsoft Corp . ) and NVivo ( QSR International ) software for qualitative data analysis .   \n",
       "877                                                                                                                                 Analysis was then done by one researcher ( SS ) using MSOffice ( Word and Excel , Microsoft Corp . ) and NVivo ( QSR International ) software for qualitative data analysis .   \n",
       "881                                                                                                                                 Analysis was then done by one researcher ( SS ) using MSOffice ( Word and Excel , Microsoft Corp . ) and NVivo ( QSR International ) software for qualitative data analysis .   \n",
       "955                                                                                                                                                                                                   The data were analyzed using SPSS software ( Statistical Package for the Social Sciences , version 17.0 ) .   \n",
       "976                                                                                                                                                                                                All three stages of analysis were conducted using SPSS software ( IBM SPSS Statistics 20 , IBM Corporation ) .   \n",
       "1069                                                                                           The images were subsequently processed and analyzed on a pixel - by - pixel basis using the Matlab Image Processing Toolbox ( Mathworks , Matlab v 2011b ) using script modified from Theyel and colleagues [20] .   \n",
       "1071                                                                                           The images were subsequently processed and analyzed on a pixel - by - pixel basis using the Matlab Image Processing Toolbox ( Mathworks , Matlab v 2011b ) using script modified from Theyel and colleagues [20] .   \n",
       "1073                                                                                           The images were subsequently processed and analyzed on a pixel - by - pixel basis using the Matlab Image Processing Toolbox ( Mathworks , Matlab v 2011b ) using script modified from Theyel and colleagues [20] .   \n",
       "1076                                                                                           The images were subsequently processed and analyzed on a pixel - by - pixel basis using the Matlab Image Processing Toolbox ( Mathworks , Matlab v 2011b ) using script modified from Theyel and colleagues [20] .   \n",
       "1139                                                                                                                              Statistical analyses were conducted using Statistical Package for the Social Sciences Version 20.0 ( SPSS 20.0 ) and Analysis of Moment Structures Version 20.0 ( AMOS 20.0 ) .   \n",
       "1154                For the pre - processing and statistical analyses , the statistical parametric mapping software package ( SPM 5 , Wellcome Department of Cognitive Neurology , London ; www.fil.ion.ucl.ac.uk/spm ) was used and implemented in Matlab ( Mathworks , Inc . , Natick , MA , USA release 14 ) .   \n",
       "1174                All analyses were performed using SAS / STAT ™ software version 9.1 ( SAS Institute Inc . , Cary , NC ) , Statistical Package for the Social Sciences ™ for Windows ™ ( SPSS ™ version 14.1 , SPSS Inc . , Chicago , IL ) , and RevMan version 4.1 ( Cochrane Collaboration , Oxford , UK ) .   \n",
       "1193                All analyses were performed using SAS / STAT ™ software version 9.1 ( SAS Institute Inc . , Cary , NC ) , Statistical Package for the Social Sciences ™ for Windows ™ ( SPSS ™ version 14.1 , SPSS Inc . , Chicago , IL ) , and RevMan version 4.1 ( Cochrane Collaboration , Oxford , UK ) .   \n",
       "1228                                                                                                                           DLMs were implemented using the dlnm package version 2.2.6 in R Studio ( Vienna , Austria ) and other analyses were performed in SAS ( v 9.4 , SAS Institute Inc . , Cary , NC ) .   \n",
       "1345                                                                                          The IVDDM and IVDGE of feedstuffs were compared using the MIXED procedure of SAS ( SAS v 9.2 , SAS Institute Inc . , Cary , NC , USA ) , where feedstuffs were treated as fixed factor and batch as random factor .   \n",
       "1351  We generated multilevel logistic regression models with random intercepts to evaluate the impact of patients and trust factors on each dependent variable using MLWin V. 2.31 software ( Rasbash J , Charlton C , Browne W , et al . MLwiN , Version 2.31 . Centre for Multilevel Modelling , University...   \n",
       "1353  We generated multilevel logistic regression models with random intercepts to evaluate the impact of patients and trust factors on each dependent variable using MLWin V. 2.31 software ( Rasbash J , Charlton C , Browne W , et al . MLwiN , Version 2.31 . Centre for Multilevel Modelling , University...   \n",
       "1357  We generated multilevel logistic regression models with random intercepts to evaluate the impact of patients and trust factors on each dependent variable using MLWin V. 2.31 software ( Rasbash J , Charlton C , Browne W , et al . MLwiN , Version 2.31 . Centre for Multilevel Modelling , University...   \n",
       "1399                                                                                                                                                           CIBERSORT analyses were performed on RNAseq data by using the analytical tool developed by Newman et al . ( http://cibersort.stanford.edu ) [30] .   \n",
       "1401                                                                                                                                                           CIBERSORT analyses were performed on RNAseq data by using the analytical tool developed by Newman et al . ( http://cibersort.stanford.edu ) [30] .   \n",
       "1421                                                                                                                                                                                                                                           Data were imported and normalized using the ChAMP R package [59] .   \n",
       "1422                                                                                                                                                                                                                                           Data were imported and normalized using the ChAMP R package [59] .   \n",
       "1438                                                                                                                                                                                     Differential expression genes between the infected and control groups were identified with the DESeq of R package [13] .   \n",
       "1439                                                                                                                                                                                     Differential expression genes between the infected and control groups were identified with the DESeq of R package [13] .   \n",
       "\n",
       "                                     subject_text target_label_binary  \\\n",
       "9                                           PyPhi           PlugIn_of   \n",
       "16                                       CoXpress           PlugIn_of   \n",
       "30                            The Mathworks Inc .        Developer_of   \n",
       "39                            The Mathworks Inc .        Developer_of   \n",
       "54                                   Bioconductor           PlugIn_of   \n",
       "94                                           CBFA           PlugIn_of   \n",
       "96                    http://www.optflux.org/cbfa              URL_of   \n",
       "106                   http://www.optflux.org/cbfa              URL_of   \n",
       "157                                     BaseTools          License_of   \n",
       "172                                     BaseTools          License_of   \n",
       "187                                          HDDM           PlugIn_of   \n",
       "351                                        gplots           PlugIn_of   \n",
       "428                                          SPSS        Developer_of   \n",
       "442                                          SPSS        Developer_of   \n",
       "447                                           SAS    Specification_of   \n",
       "473                                           SMI        Developer_of   \n",
       "523                                   [ 61 , 66 ]         Citation_of   \n",
       "578                               SPSS Statistics    Specification_of   \n",
       "614                                          [53]         Citation_of   \n",
       "615                                          [53]         Citation_of   \n",
       "617                                   [ 62 , 63 ]         Citation_of   \n",
       "624                                          [58]         Citation_of   \n",
       "625                                          [59]         Citation_of   \n",
       "630                                          [58]         Citation_of   \n",
       "631                                          [59]         Citation_of   \n",
       "671                               SPSS Statistics    Specification_of   \n",
       "719                 http://www.matrixscience.com/              URL_of   \n",
       "720                 http://www.matrixscience.com/              URL_of   \n",
       "723                             www.r-project.org              URL_of   \n",
       "724                             www.r-project.org              URL_of   \n",
       "730                          http://www.nltk.org/              URL_of   \n",
       "731                          http://www.nltk.org/              URL_of   \n",
       "861                                          [38]         Citation_of   \n",
       "862                                          [38]         Citation_of   \n",
       "867                              Microsoft Corp .        Developer_of   \n",
       "873                              Microsoft Corp .        Developer_of   \n",
       "877                              Microsoft Corp .        Developer_of   \n",
       "881                              Microsoft Corp .        Developer_of   \n",
       "955   Statistical Package for the Social Sciences    Specification_of   \n",
       "976                               SPSS Statistics    Specification_of   \n",
       "1069                                    Mathworks        Developer_of   \n",
       "1071                                       Matlab    Specification_of   \n",
       "1073                                    Mathworks        Developer_of   \n",
       "1076                                    Mathworks        Developer_of   \n",
       "1139                                         AMOS    Specification_of   \n",
       "1154                                          SPM    Specification_of   \n",
       "1174                          SAS Institute Inc .        Developer_of   \n",
       "1193                          SAS Institute Inc .        Developer_of   \n",
       "1228                                         dlnm           PlugIn_of   \n",
       "1345                                          SAS    Specification_of   \n",
       "1351  Rasbash J , Charlton C , Browne W , et al .        Developer_of   \n",
       "1353                                        MLwiN    Specification_of   \n",
       "1357  Rasbash J , Charlton C , Browne W , et al .        Developer_of   \n",
       "1399                http://cibersort.stanford.edu              URL_of   \n",
       "1401                http://cibersort.stanford.edu              URL_of   \n",
       "1421                                         [59]         Citation_of   \n",
       "1422                                         [59]         Citation_of   \n",
       "1438                                         [13]         Citation_of   \n",
       "1439                                         [13]         Citation_of   \n",
       "\n",
       "                                      object_text          relation  \\\n",
       "9                                          GitHub               nil   \n",
       "16                                              R               nil   \n",
       "30                                         Matlab      Developer_of   \n",
       "39                       Image Processing Toolbox               nil   \n",
       "54                                       FunciSNP               nil   \n",
       "94                                        Optflux               nil   \n",
       "96                                           CBFA            URL_of   \n",
       "106                                       Optflux               nil   \n",
       "157                             SSPACE - LongRead        License_of   \n",
       "172                                         BLASR               nil   \n",
       "187                                        Python               nil   \n",
       "351                                             R         PlugIn_of   \n",
       "428        Statistical Package for Social Science      Developer_of   \n",
       "442                                           SAS               nil   \n",
       "447                                           SAS  Specification_of   \n",
       "473                                       iView X      Developer_of   \n",
       "523                      articulatory synthesizer       Citation_of   \n",
       "578                                          SPSS  Specification_of   \n",
       "614                                          clam       Citation_of   \n",
       "615                                             R               nil   \n",
       "617                                         rioja       Citation_of   \n",
       "624                             adjustedRandIndex       Citation_of   \n",
       "625                             adjustedRandIndex       Citation_of   \n",
       "630                                             R               nil   \n",
       "631                                             R               nil   \n",
       "671                                          SPSS  Specification_of   \n",
       "719                                        MASCOT            URL_of   \n",
       "720                                Matrix Science               nil   \n",
       "723                                             R            URL_of   \n",
       "724                       R Development Core Team               nil   \n",
       "730                                        Python            URL_of   \n",
       "731                                          NLTK               nil   \n",
       "861                                       PROCESS       Citation_of   \n",
       "862                                          SPSS               nil   \n",
       "867                                      MSOffice      Developer_of   \n",
       "873                                          Word      Developer_of   \n",
       "877                                         Excel      Developer_of   \n",
       "881                                         NVivo               nil   \n",
       "955                                          SPSS  Specification_of   \n",
       "976                                          SPSS  Specification_of   \n",
       "1069                                       Matlab      Developer_of   \n",
       "1071                                       Matlab  Specification_of   \n",
       "1073                     Image Processing Toolbox      Developer_of   \n",
       "1076                                       Matlab               nil   \n",
       "1139                Analysis of Moment Structures  Specification_of   \n",
       "1154               statistical parametric mapping  Specification_of   \n",
       "1174                                          SAS      Developer_of   \n",
       "1193  Statistical Package for the Social Sciences               nil   \n",
       "1228                                     R Studio               nil   \n",
       "1345                                          SAS  Specification_of   \n",
       "1351                                        MLWin      Developer_of   \n",
       "1353                                        MLWin  Specification_of   \n",
       "1357                                        MLwiN               nil   \n",
       "1399                                    CIBERSORT            URL_of   \n",
       "1401                               Newman et al .               nil   \n",
       "1421                                        ChAMP       Citation_of   \n",
       "1422                                            R               nil   \n",
       "1438                                        DESeq       Citation_of   \n",
       "1439                                            R               nil   \n",
       "\n",
       "              pred  \n",
       "9        PlugIn_of  \n",
       "16       PlugIn_of  \n",
       "30             nil  \n",
       "39    Developer_of  \n",
       "54       PlugIn_of  \n",
       "94       PlugIn_of  \n",
       "96             nil  \n",
       "106            nil  \n",
       "157            nil  \n",
       "172            nil  \n",
       "187      PlugIn_of  \n",
       "351            nil  \n",
       "428            nil  \n",
       "442            nil  \n",
       "447            nil  \n",
       "473            nil  \n",
       "523            nil  \n",
       "578            nil  \n",
       "614            nil  \n",
       "615    Citation_of  \n",
       "617            nil  \n",
       "624            nil  \n",
       "625            nil  \n",
       "630    Citation_of  \n",
       "631    Citation_of  \n",
       "671            nil  \n",
       "719            nil  \n",
       "720         URL_of  \n",
       "723            nil  \n",
       "724         URL_of  \n",
       "730            nil  \n",
       "731         URL_of  \n",
       "861            nil  \n",
       "862    Citation_of  \n",
       "867            nil  \n",
       "873            nil  \n",
       "877            nil  \n",
       "881   Developer_of  \n",
       "955            nil  \n",
       "976            nil  \n",
       "1069           nil  \n",
       "1071           nil  \n",
       "1073           nil  \n",
       "1076  Developer_of  \n",
       "1139           nil  \n",
       "1154           nil  \n",
       "1174           nil  \n",
       "1193  Developer_of  \n",
       "1228     PlugIn_of  \n",
       "1345           nil  \n",
       "1351           nil  \n",
       "1353           nil  \n",
       "1357  Developer_of  \n",
       "1399           nil  \n",
       "1401           nil  \n",
       "1421           nil  \n",
       "1422   Citation_of  \n",
       "1438           nil  \n",
       "1439   Citation_of  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ha"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c14de626-a316-47fb-8ffd-3bd528308a34",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85cdd9e2-03e6-4882-ac9d-b948fa91ee51",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "gpuType": "T4",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
